cpp/src/cores/trim.cu

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+
+#include <cugraph/algorithms.hpp>
+#include <cugraph/graph_functions.hpp>
+#include <cugraph/graph_view.hpp>
+#include <cugraph/utilities/error.hpp>
+
+#include <thrust/copy.h>
+#include <thrust/iterator/discard_iterator.h>
+
+namespace cugraph {
+
+template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+std::tuple<rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<vertex_t>,
+           std::optional<rmm::device_uvector<weight_t>>>
+trim(raft::handle_t const& handle,
+       graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view,
+       bool do_expensive_check)
+{
+
+  
+  auto in_degree = graph_view.compute_in_degree();
+  auto out_degree = graph_view.compute_out_degree();
+ 
+  rmm::device_uvector<vertex_t> remaining_vertices(graph_view.local_vertex_partition_range_size(),  handle.get_stream());
+
+  remaining_vertices.resize(
+    thrust::distance( remaining_vertices.begin(), thrust::copy_if(
+        handle.get_thrust_policy(), thrust::make_counting_iterator(graph_view.local_vertex_partition_range_first()),
+        thrust::make_counting_iterator(graph_view.local_vertex_partition_range_last()),  remaining_vertices.begin(),
+        [ in_degree ] __device__(
+          auto in_degree) { return in_degree <= 1; })),
+    handle.get_stream());
+
+  return remaining_vertices;
```

Comment:
Finding vertices with non-zero in|out degrees and finding induced subgraph is one way to solve this problem.

Another more direct approach is to use extract_if_e.

```

  if (graph_view.count_self_loops(handle) > edge_t{0}) {
    auto [srcs, dsts] = extract_if_e(handle,
                                     graph_view,
                                     edge_src_dummy_property_t{}.view(),
                                     edge_dst_dummy_property_t{}.view(),
                                     is_not_self_loop_t<vertex_t>{});

    if constexpr (multi_gpu) {
      std::tie(srcs, dsts, std::ignore, std::ignore) =
        detail::shuffle_ext_vertex_pairs_to_local_gpu_by_edge_partitioning<vertex_t,
                                                                           edge_t,
                                                                           weight_t,
                                                                           int32_t>(
          handle, std::move(srcs), std::move(dsts), std::nullopt, std::nullopt);
    }

    std::tie(*modified_graph, std::ignore, std::ignore, renumber_map) =
      create_graph_from_edgelist<vertex_t, edge_t, weight_t, int32_t, false, multi_gpu>(
        handle,
        std::nullopt,
        std::move(srcs),
        std::move(dsts),
        std::nullopt,
        std::nullopt,
        cugraph::graph_properties_t{true, graph_view.is_multigraph()},
        true);

    modified_graph_view = (*modified_graph).view();
  }
```
https://github.com/rapidsai/cugraph/blob/branch-23.04/cpp/src/community/triangle_count_impl.cuh#L201

Here, instead of passing dummy property views, you need to pass in_degrees and out_degrees.

See
```
    edge_src_property_t<decltype(cur_graph_view), uint8_t> edge_src_in_two_cores(handle,
                                                                                 cur_graph_view);
    edge_dst_property_t<decltype(cur_graph_view), uint8_t> edge_dst_in_two_cores(handle,
                                                                                 cur_graph_view);
```
https://github.com/rapidsai/cugraph/blob/branch-23.04/cpp/src/community/triangle_count_impl.cuh#L246

and

```
    update_edge_src_property(
      handle, cur_graph_view, in_two_core_flags.begin(), edge_src_in_two_cores);
    update_edge_dst_property(
      handle, cur_graph_view, in_two_core_flags.begin(), edge_dst_in_two_cores);
```
https://github.com/rapidsai/cugraph/blob/branch-23.04/cpp/src/community/triangle_count_impl.cuh#L257

and

```
                                     edge_src_in_two_cores.view(),
                                     edge_dst_in_two_cores.view(),
```
https://github.com/rapidsai/cugraph/blob/branch-23.04/cpp/src/community/triangle_count_impl.cuh#L263

to figure out how to do this.

Let me know if you have additional questions.